### DISCUSSION TITLE
Google Photos API updates are coming

### DISCUSSION THREAD

**Discussion by siburny** [mtalcott/google-photos-deduper#65] [General]
I just received an email from Google saying that they will remove `photoslibrary` scope from the API (March 2025). Essentially, it will not allow us to download a full photo library using this API, which will kill this app. Did anyone else get that? What are our options?

Source: https://developers.google.com/photos/support/updates#affected-scopes-methods

**Comment by aleon1220** [comment]
I did read the api docs announcement. I guess running batches per year could work out but requires some refactoring?
any thoughts on the lifecycle of your open-source project?

**Comment by mtalcott** [comment]
I also received the update. It is unfortunate and yes, the tool will need some changes to keep working past March 2025. Here's my latest thinking on approaches:

- **Option 1**: Use the new [Google Photos Picker API](https://developers.google.com/photos/picker/guides/get-started-picker).
    - **Option 1a**: I am hoping it provides a way for users to [easily select all photos in their library](https://www.youtube.com/watch?v=Yo7f1NtyYe0). If so, switch to using its [OAuth scope](https://developers.google.com/photos/picker/reference/rest/v1/sessions/create#authorization-scopes), [create a session](https://developers.google.com/photos/picker/reference/rest/v1/sessions/create), and [pull the selected media items](https://developers.google.com/photos/picker/reference/rest/v1/mediaItems/list) for processing. Relaunch a new session as necessary if it expires.
    - **Option 1b**: If there's no easy way to select all, hey, I already have a [Chrome extension](https://github.com/mtalcott/google-photos-deduper/blob/main/chrome_extension/README.md) to utilize to automate the task to select all. Then proceed same as 1a.
- **Option 2**: Use Google Photos' undocumented (`batchexecute`) Web API, perhaps via [xob0t/Google-Photos-Toolkit's GPTK API](https://github.com/xob0t/Google-Photos-Toolkit/blob/main/README.md#use-gptks-api), which I just learned about through a [Reddit post](https://www.reddit.com/r/googlephotos/comments/1g0y8n6/comment/lrdb0ur/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).
     - This would likely require shifting most of the media item listing & thumbnail downloading logic into the browser from Python.
     - Come to think of it, the MediaPipe (now Google AI Edge) tool I'm already using to detect duplicates [has a JS client that can also be used for image embedding & image similarity calculation](https://ai.google.dev/edge/mediapipe/solutions/vision/image_embedder/web_js)... This means **the entire tool could potentially be bundled into Chrome extension**, which would greatly simplify setup and reach a wider audience.
     - Utilizing the unofficial API also doesn't feel any more risky given that Google clearly makes major breaking changes to its _official_ APIs as well, and the fact that the Chrome Extension is already dependent an unstable interface (the Google Photos UI) to automate deletion. In fact, the unofficial API is probably more stable than the UI the tool is currently based on.

Leaning toward Option 2.

**Reply by siburny** [reply]
Option 1 will work only for a small library/batch. I installed an example Picker API app, tried to select 5000 images, and it failed with an error (it worked with about 500 images selected though).

**Reply by mtalcott** [reply]
Good to know - thanks for trying it out and reporting back

**Reply by mtalcott** [reply]
I began work on Option 2 on a branch [here](https://github.com/mtalcott/google-photos-deduper/compare/main...full_chrome_ext). I ran into an issue where the GPTK API only returns a small portion of the total media items, and haven't been able to work around it. Unfortunately, finding time to work on the project has been difficult, and I don't anticipate that changing anytime soon.

**Reply by aleon1220** [reply]
thanks for your efforts. I was able to use your app and got rid of many duplicates. Google photos is behaving strangely and now i am going manual 5 mins per day as my collection requires a lot of manual interaction. 
thanks so much

**Reply by xob0t** [reply]
Hello!
GPTK's dev here.
I've recently added a Similarity filter. It groups similar images together, leaves out images without pairs.
Works by comparing difference hashes.
The best way to use it if you want to find duplicates - set threshold to 0.95, and use "add to album" action.


> GPTK API only returns a small portion of the total media items

Have not encountered this, but if you have any other issues or questions, let me know!

**Reply by mtalcott** [reply]
Hi @xob0t! That's great to hear there might be a way for folks to leverage GPTK for duplicate management.

I've come close to reaching out a few times on the issue, but I wanted to get reliable repro steps first. It only occurred with my primary personal account with ~60k mediaItems. Then my available time to focus on this project dropped from little to zero, and I haven't had a chance to circle back. If I can find some time I'll put some effort in and reach out directly!

**Reply by xob0t** [reply]
> It only occurred with my primary personal account with ~60k mediaItems

There is a possibility that was a bug that was recently fixed.

> That's great to hear there might be a way for folks to leverage GPTK for duplicate management.

For now, it is far from perfect :)
Works okay for small batches of media, but hashing and matching thousands of them can take hours.
Due to GP's security policy, and GPTK being a userscript, webworkers are not allowed, all the work runs on a single thread.
It is tempting to implement webworkers, and release it as extension. Maybe rework the UI while at it- it is bothering me for a while now.
But that is quite a lot of work, maybe someday.

**Comment by ajtazer** [comment]
> this tool will stop working on March 31, 2025

😭 why am i always late
_today's date is 7April2025_

**Reply by crazy-treyn** [reply]
Same bro..

**Comment by ZonD80** [comment]
I have made an extension for Chrome that does the same. It's partly open-sourced and freemium. If repository owner grants me permission to mention it, I will be happy to share discount code and share some revenue with him generated by it

**Reply by aleon1220** [reply]
do you offer a trial?
https://gpdrm.com/

**Reply by ZonD80** [reply]
Yes. Free version is always available with limited features

**Reply by anthonyjclarke** [reply]
Its very good!

**Reply by aleon1220** [reply]
i think i might buy it. 
I am working on my personal project of making my memories digital. Do you think is best to manually sort out the dates first and run the dedup?

**Reply by ZonD80** [reply]
It is integrated with search bar, so you can use any search term

**Comment by mian196** [comment]
Any update?

**Comment by mtalcott** [comment]
Hi all - new version is available that bypasses the restrictions in Google Photos API by leveraging Google Photos' undocumented (`batchexecute`) Web API via [xob0t/Google-Photos-Toolkit's GPTK API](https://github.com/xob0t/Google-Photos-Toolkit). It's much easier to install and run - just download the Chrome extension rather than complex Docker & Google OAuth setup!
